### PR TITLE
"Don't directly throw Exception" System.Threading.Tasks.Dataflow

### PR DESCRIPTION
**Pull request by MarcoRossignoli**
Contributes to #23748 .

cc: @kouvel, @tarekgh

### CODE REVIEW

File: src/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.cs [hunk 1]
```diff
@@ -1353,7 +1353,7 @@ private void CleanupAndComplete(ReceiveCoreByLinkingCleanupReason reason)
                         {
                             // Complete with the received exception
                             var target = (ReceiveTarget<T>)state;
-                            try { target.TrySetException(target._receivedException ?? new Exception()); }
+                            try { target.TrySetException(target._receivedException); }
```
Comment (MarcoRossignoli) [inline src/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.cs]: Is there a reason to throw generic Exception here? Above there is an assertion " Debug.Assert(_receivedException != null, "We need an exception with which to fault the task.");" so _receivedException should never be null. I also try to simulate a throw Exception() here and the exception bubble out to user code. What do you think?

**Comment by MarcoRossignoli** [comment]
@dotnet-bot test NETFX x86 Release Build

File: src/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.cs (hunk 1 above)
Comment (stephentoub) [reply src/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.cs]: Do we know that the assert is in fact correct, that there's no way _receivedException could be null?  Cleaning up the code is nice, but the `?? new Exception()` really isn't hurting anything, and if there's even the remote chance _receivedException could be null here, leaving the fallback in place is necessary.  If you can prove that _receivedException can never be null here, the fix looks good.

Comment (MarcoRossignoli) [reply src/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.cs]: I get the point, i'll try to understand if _receivedException can never be null, also another idea could be try/catch and rethrow argumentnull with a more clear errore message, so in remote chance of null the user get a more useful informativo, What do you think?

**Comment by MarcoRossignoli** [comment]
@stephentoub i've been looking over the code and seems that target._receivedException is never null: <br/>
ReceiveCoreByLinkingCleanupReason.ErrorDuringCleanup is used only on switch instead of ReceiveCoreByLinkingCleanupReason.SourceProtocolError [here](https://github.com/MarcoRossignoli/corefx/blob/4483ff112fe8f741795c517a44e03553584f9136/src/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.cs#L1350),<br/>
ReceiveCoreByLinkingCleanupReason.SourceProtocolError is always used after assignment of target._receivedException to an exception in 4 place [1](https://github.com/MarcoRossignoli/corefx/blob/4483ff112fe8f741795c517a44e03553584f9136/src/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.cs#L1345),[2](https://github.com/MarcoRossignoli/corefx/blob/4483ff112fe8f741795c517a44e03553584f9136/src/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.cs#L1348),[3](https://github.com/MarcoRossignoli/corefx/blob/4483ff112fe8f741795c517a44e03553584f9136/src/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.cs#L1279),[4](https://github.com/MarcoRossignoli/corefx/blob/4483ff112fe8f741795c517a44e03553584f9136/src/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.cs#L1117).<br/>
The only other place is [here](https://github.com/MarcoRossignoli/corefx/blob/4483ff112fe8f741795c517a44e03553584f9136/src/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.cs#L1220) where the exception is set some lines above(es: TestReceive_FaultySourceConsume).<br/>
There is no code set target._receivedException == null. The only risk is to eat an exception on race.<br/>
If you think this change is too risky(unobserved exceptions and never set tcs) we could only change Exception with ArgumentNull or InvalidOperation with a more clear bug message es. "Internal error".

File: src/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.cs (hunk 1 above)
Comment (MarcoRossignoli) [reply src/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.cs]: @tarekgh help me to understand, i don't found the right code path. In method "OfferMessage" "status" is local and is assigned to Accepted or DecliningPermanently. DecliningPermanently is assigned only inside catch block where _receivedException is set. How do the code flow to  CleanupAndComplete(ReceiveCoreByLinkingCleanupReason.SourceProtocolError); with _receivedException null?You say "ReceiveCoreByLinkingCleanupReason.ErrorDuringCleanup which not used yet", but i test that is used as a fallback for ReceiveCoreByLinkingCleanupReason.SourceProtocolError, what am I missing?Thanks in advance.

**Comment by tarekgh** [comment]
@MarcoRossignoli you are right, I missed the status initialization. thank you.

I am still thinking removing the assert and keeping the original code is better here. the reason is, if ReceiveCoreByLinkingCleanupReason.ErrorDuringCleanup is used in the future code we'll have to ensure setting the _receivedException which I am not sure if this would be mandatory. The same applies if we use ReceiveCoreByLinkingCleanupReason.SourceProtocolError in other places too in the future code.

**Comment by MarcoRossignoli** [comment]
@tarekgh i understand your concern about future use. My last thought, the goal of this issue is to improve the "exception experience" for users when exception raise following the " Framework Design Guidelines". If you think could be useful we can throw an InvalidException(or another one) with a message like "Task fail" or something like that. If you think this is not so useful e remove the assert and we can close this PR.

**Comment by tarekgh** [comment]
I like the idea to have the exception have an explicit helpful message. looking at 

```C#
            /// <summary>An error occurred while linking up the target.</summary>
            SourceProtocolError = 4,
            /// <summary>An error during cleanup after completion for another reason.</summary>
            ErrorDuringCleanup = 5
```

we can use the comments on these definitions in the exception message. 
Also, throwing InvalidOperationException is reasonable to me.

**Comment by MarcoRossignoli** [comment]
@tarekgh done

File: src/System.Threading.Tasks.Dataflow/src/Resources/Strings.resx [hunk 2]
```diff
@@ -97,4 +97,7 @@
   <data name="ConcurrentCollection_SyncRoot_NotSupported" xml:space="preserve">
     <value>The SyncRoot property may not be used for the synchronization of concurrent collections.</value>
   </data>
+  <data name="InvalidOperation_ErrorDuringCleanup" xml:space="preserve">
+    <value>An error during cleanup after completion for another reason.</value>
```
Comment (tarekgh) [inline src/System.Threading.Tasks.Dataflow/src/Resources/Strings.resx]: >An error during cleanup after completion for another reason. [](start = 11, length = 60)

We may write this as "An error occurred while linking up the target or during the cleanup after completion"

Comment (MarcoRossignoli) [reply src/System.Threading.Tasks.Dataflow/src/Resources/Strings.resx]: done

**Comment by tarekgh** [comment]
LGTM. Thanks @MarcoRossignoli

**Pull request merged**